src/protocol/codecs/tagmsg.rs

```diff
@@ -0,0 +1,151 @@
+
+        Ok(match tag {
+            "??" => Self::UnknownMsg,
+            "AV" => Self::AgreementVote,
+            "MI" => Self::MsgOfInterest,
+            "MS" => Self::MsgDigestSkip,
+            "NP" => Self::NetPrioResponse,
+            "pi" => Self::Ping,
+            "pj" => Self::PingReply,
+            "PP" => Self::ProposalPayload,
+            "SP" => Self::StateProofSig,
+            "TS" => Self::TopicMsgResp,
+            "TX" => Self::Txn,
+            "UC" => Self::UniCatchupReq,
+            "UE" => Self::UniEnsBlockReq,
+            "VB" => Self::VoteBundle,
+            _ => return Err(invalid_data!("unexpected tag")),
+        })
+    }
+}
+
+#[derive(Debug)]
+pub enum Payload {
+    MsgOfInterest(MsgOfInterest),
+    NotImplemented,
+}
+
+/// [MsgOfInterest] contains a tag list in which the node is interested.
+#[derive(Debug)]
+pub struct MsgOfInterest {
+    /// Message tags for which the node is interested (subscribed).
+    // TODO(Rqnsom): convert to a hashset.
+    pub tags: String,
+}
+
+impl MsgOfInterest {
+    /// Constructs a [MsgOfInterest] message from a [input] byte stream using the [topic] codec.
+    fn construct(input: &mut BytesMut, topic: &mut TopicCodec) -> Result<Self, io::Error> {
+        let mut topics = topic
+            .decode(input)?
+            .ok_or_else(|| invalid_data!("couldn't decode the topics"))?;
+        if topics.len() != 1 {
+            return Err(invalid_data!("expected a single topic"));
+        }
+
+        let tag_topic = topics.pop().unwrap();
+        if tag_topic.key != "tags" {
+            return Err(invalid_data!("expected 'tags' topic"));
+        }
+
+        Ok(Self {
+            tags: tag_topic.value,
+        })
+    }
+}
+
+/// [TagMessage] represents sets of messages received within the data field of a WebSocket message.
+///
+/// Different messages (identified by tags) use different encoding. Two encoding schemes
+/// are used:
+///  - Topic encoding (key-value string pairs)
+///  - MessagePack encoding
+#[derive(Debug)]
+pub struct TagMessage {
+    /// Tag represents a message type identifier.
+    pub tag: Tag,
+
+    /// Each tag has its respective payload data.
+    pub payload: Payload,
+}
+
+/// [TagMsgCodec] is the codec for [TagMessage] messages.
+#[derive(Clone)]
+pub struct TagMsgCodec {
+    /// The associated node's span.
+    span: Span,
+
+    /// Codec for topics which are key-value string pairs.
+    topic: TopicCodec,
+}
+
+impl TagMsgCodec {
+    pub fn new(span: Span) -> Self {
+        Self {
+            topic: TopicCodec::default(),
+            span,
+        }
+    }
+}
+
+impl Decoder for TagMsgCodec {
+    type Item = TagMessage;
+    type Error = io::Error;
+
+    fn decode(&mut self, src: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
+        const TAG_LEN: usize = 2;
+
+        if src.len() < TAG_LEN {
+            return Err(invalid_data!("invalid tag length"));
+        }
```

Comment:
it's probably not possible to get a truncated message from the outer codec, so I guess we don't need to worry about it here :detective: